### PR TITLE
repo-updater: Refactor Store scanning and pagination

### PR DESCRIPTION
**Pull request by tsenart**
In preparation for extending `Store` with `ListSources` and
`UpsertSources`, which is required for upcoming migrations, this
commit refactors internal methods of pagination and scanning to leverage
closure side effects for polymorphism.

This is a more type-safe approach than refactoring to use `*[]interface{}`
instead of `*[]*Repo`.

Part of #2025

### CODE REVIEW

File: cmd/repo-updater/repos/store.go [hunk 1]
```diff
@@ -96,13 +96,20 @@ func (s *DBStore) Done(errs ...*error) {
...
+	_, err := s.list(ctx, getRepoByNameQuery(name), func(s scanner) (int64, error) {
+		err := scanRepo(&r, s)
+		return int64(r.ID), err
+	})
+
+	switch {
+	case err != nil:
 		return nil, err
-	} else if len(repos) == 0 {
+	case r.ID == 0:
```
Comment (keegancsmith) [inline cmd/repo-updater/repos/store.go]: so sequences in Postgres 9.4+ always start at 1 by default?

Comment (tsenart) [reply cmd/repo-updater/repos/store.go]: Yes, [that's the case](https://www.postgresql.org/docs/9.6/sql-createsequence.html), but it does seem a little surprising at first sight. So I'll change this to instantiate a Repo with a marker id of -1 and compare against that instead.

**Comment by tsenart** [comment]
Please take another look.

**Comment by keegancsmith** [comment]
LGTM

**Pull request merged**